app/workers/publishing_api_legacy_tags_worker.rb

```diff
@@ -0,0 +1,72 @@
+class PublishingApiLegacyTagsWorker
+  include Sidekiq::Worker
+  API = Services.publishing_api
+
+  def perform(model_id, model_class)
+    model = model_class.constantize.find(model_id)
+    previous_version, taxon_ids = get_taxon_ids(model)
+    parent_ids = taxon_ids.flat_map { |id| expand_parents(id) }
+    legacy_ids = get_legacy_links(taxon_ids + parent_ids)
+
+    links = { policy_areas: [], policies: [], topics: [] }
+    update_topics(links, legacy_ids, model)
+    update_policies(links, legacy_ids, model)
+    update_policy_areas(links, legacy_ids, model)
+
+    params = { previous_version: previous_version, links: links }
+    API.patch_links(model.content_id, params)
+  end
+
+private
+
+  def update_topics(links, legacy_ids, model)
+    return unless model.respond_to?(:specialist_sector_tags)
+    topics = API.get_linkables(document_type: 'topic')
+      .to_a.map { |topic| topic["content_id"] } & legacy_ids
+
+    links[:topics] += topics
+    model.primary_specialist_sector_tag = topics.first
+    model.secondary_specialist_sector_tags = topics.drop(1)
+  end
+
+  def update_policies(links, legacy_ids, model)
+    return unless model.respond_to?(:policy_content_ids)
+    policies = Policy.all.map(&:content_id) & legacy_ids
+    links[:policies] += policies
+
+    model.policy_content_ids = policies
+
+    parent_policies = model.policy_content_ids - policies
+    links[:policy_areas] += parent_policies
```

Comment:
So, after much faffing, I've come to the opinion that this isn't quite right.

The assignment to `model.policy_content_ids` can silently add additional `content_id`s to the associated `policy_content_ids`. However, I don't think this will happen in the live environment, and I don't think the additional `content_id`s are actually policy areas, but instead will be other policies.

Taking the last point first, policies are published from Policy Publisher, and it does set the `policy_areas` links [1]. What it sets them to are not policy areas though, but instead other policies, which it refers to as parent policies (or just policies).

In reality though, there are not any sub policies. I'm not sure if this is intentional or not... all I know is that from Jan 2016 until Jan 2018, any attempt to edit a sub policy in Policy Publisher will have caused it to become a policy, and loose the parent information [2].

So, all in all. We may as well try and still cope with this behaviour, but I think this should look more like:
```ruby
model.policy_content_ids = policies
# It is possible that policies get added on assignment, so keep this behaviour here
links[:policies] = model.policy_content_ids
```

Now, there is a separate issue around the mapping from policies to policy areas that is maintained within Whitehall, as it uses that to add policy areas through JavaScript on the edit form. I'm still undecided what to do with that...

1: https://github.com/alphagov/policy-publisher/blob/master/app/presenters/links_presenter.rb#L17
2: https://github.com/alphagov/policy-publisher/pull/214

Reply:
Yuk. When I saw "policy_areas" in related_policies.rb I assumed it meant what it said. Lesson learnt, I guess - thanks for finding that out :-)

Your suggestion make sense, plus also re-using the logic from the admin/policies_controller.rb to do the mapping from policies to policy areas.

As the links hash would then be populated just once for each kind of link, it also means we can use the LinkPresenter like in the old-to-new code.

Reply:
I've put together a patch [1] to remove the use of the policy_areas link type from Policy Publisher. Once that is gone, maybe something can be done about the Whitehall side of this functionality as well.

1: https://github.com/alphagov/policy-publisher/pull/255